base/base/LineReader.cpp

```diff
@@ -36,10 +36,9 @@ bool hasInputData()
 
 }
 
-std::optional<LineReader::Suggest::WordsRange> LineReader::Suggest::getCompletions(const String & prefix, size_t prefix_length) const
+LineReader::Suggest::Words LineReader::Suggest::getCompletions(const String & prefix, size_t prefix_length)
 {
-    if (!ready)
-        return std::nullopt;
+    std::lock_guard lock(mutex);
```

Comment:
why this mutex is now needed? Looks unrelated to adding periodical update, to something else?

Reply:
Completion is reloaded in a separate thread, and so to avoid race between read of `words_no_case`/`words` there is a mutex now here and here - https://github.com/ClickHouse/ClickHouse/pull/32370/files#diff-11fe0247064a69cca4819453e93874d9d73dff4feb242daee35dcf7e54f48f7aR168-R172